Source/WebCore/rendering/RenderBox.cpp

```diff
@@ -3720,6 +3720,27 @@ LayoutUnit RenderBox::constrainBlockMarginInAvailableSpaceOrTrim(const RenderBox
     return marginSide == MarginTrimType::BlockStart ? minimumValueForLength(style().marginBeforeUsing(&containingBlock.style()), availableSpace) : minimumValueForLength(style().marginAfterUsing(&containingBlock.style()), availableSpace);
 }
 
+bool RenderBox::hasTrimmedMargin(MarginTrimType marginTrimType) const
+{
+    ASSERT(!needsLayout());
+    auto* container = containingBlock();
+
+    // FIXME: Remove these asserts as they are implemented for each layout context
+    if (isFlexItem() || isGridItem())
+        ASSERT_NOT_IMPLEMENTED_YET();
+    else if (container->isBlockContainer())
+        ASSERT(marginTrimType == MarginTrimType::BlockStart);
+    else
+        return false;
+
+    auto isAdjacentToContainingBlockEdge = [&]() {
+        return container->isBlockContainer() && container->logicalMarginBoxTopForChild(*this) == container->borderAndPaddingBefore(); 
```

Comment:
does this actually match with your trimming logic? e.g. what happens when there are multiple zero height block boxes piled up at block-start. I assume they all get trimmed, just checking)...or when the block box is relatively positioned.

Reply:
Modified the `block-container-block-start` to contain the first scenario you mentioned. Also a couple of tests were added for the relative positioning. I think the relative positioning test cases match the spec but let me know if you think otherwise (relative positioned object will get block-start trimmed if it is a block-level first child or participates in margin collapsing on the block-start side)